### PR TITLE
[Thread]Move the encoding of the thread diagnostics attributes to the app layer

### PR DESCRIPTION
**Pull request by jmartinez-silabs**
This PR moves the `WriteThreadNetworkDiagnosticAttributeToTlv` functions to a single location for all thread devices.
It also, more importantly, removes it from the device layer/platform. 
It is compiled with the cluster server implementation when that cluster is enabled.

This is the first step to remove the app layer dependency in `src/platform/Openthread`.

The next phase will be targeted to the network commissioning cluster dependency.

### CODE REVIEW

**Review by lpbeliveau — APPROVED** [review]

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+
+/*
+ * @brief Get runtime value from the thread network based on the given attribute ID.
+ *        The info is encoded via the AttributeValueEncoder.
+ *
+ * @param attributeId Id of the attribute for the requested info.
+ * @param aEncoder Encoder to encode the attribute value.
+ *
+ * @return CHIP_NO_ERROR = Succes.
+ *         CHIP_ERROR_NOT_IMPLEMENTED = Runtime value for this attribute to yet available to send as reply
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:49]:
```suggestion
 *         CHIP_ERROR_NOT_IMPLEMENTED = Runtime value for this attribute not yet available to send as reply
```

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
+ *                                      Use standard read.
+ *         CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE = Is not a Runtime readable attribute. Use standard read
+ *         All other errors should be treated as a read error and reported as such.
+ *
+ * @note This function implementation can compile in 3 different outcome
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:54]:
```suggestion
 * @note This function implementation can compile in 3 different outcomes
```

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+ *       (1) CHIP_DEVICE_CONFIG_ENABLE_THREAD && !CHIP_DEVICE_CONFIG_USES_OTBR_POSIX_DBUS_STACK:
+ *           - Generic implementation fetching the valid thread network data from the thread stack and encoding it respectively to
+ *             the attributeID received.
+ *       (2) CHIP_DEVICE_CONFIG_ENABLE_THREAD && CHIP_DEVICE_CONFIG_USES_OTBR_POSIX_DBUS_STACK:
+ *           - Encode a NULL value for nullable attributes or 0 for the others.
+ *           - Devices using the ot-br-posix dbus stack have not yet provided the API the fetch the needed thread informations.
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:60]:
```suggestion
 *           - Devices using the ot-br-posix dbus stack have not yet provided the API to fetch the needed thread informations.
```

File: src/platform/BUILD.gn [hunk 4]
```diff
@@ -111,11 +111,22 @@ if (chip_device_platform != "none" && chip_device_platform != "external") {
         chip_enable_wifi && chip_device_platform != "darwin"
     chip_stack_lock_tracking_log = chip_stack_lock_tracking != "none"
     chip_stack_lock_tracking_fatal = chip_stack_lock_tracking == "fatal"
+
+    # This is used to identify which platforms implement their ThreadStackManager
+    # with the otbr posix dbus api.
+    chip_device_uses_otbr_posix_dbus = false
```
Comment (andy31415) [inline src/platform/BUILD.gn:117]: should this be:

```
chip_device_uses_otbr_posix_dbus = (
    chip_enable_openthread &&
        (chip_device_platform == "linux" || chip_device_platform == "tizen" ||
         chip_device_platform == "webos")
)
```

?

**Review by andy31415 — APPROVED** [review]

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 5]
```diff
@@ -0,0 +1,861 @@
...
+ *       (3) None of the conditions above
+ *           - returns CHIP_ERROR_NOT_IMPLEMENTED.
+ */
+CHIP_ERROR WriteThreadNetworkDiagnosticAttributeToTlv(AttributeId attributeId, app::AttributeValueEncoder & encoder)
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:64]: This function could have been simplified by having three sub functions that can compiled in or out based on the define values. having all 3 outcomes in one function makes it pretty huge.

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 6]
```diff
@@ -0,0 +1,861 @@
...
+{
+    CHIP_ERROR err;
+#if (CHIP_DEVICE_CONFIG_ENABLE_THREAD && !CHIP_DEVICE_CONFIG_USES_OTBR_POSIX_DBUS_STACK)
+    otInstance * otInst = ThreadStackMgrImpl().OTInstance();
+
+    if (!otDatasetIsCommissioned(otInst))
+    {
+        // For the following nullable attributes of the cluster, encodeNull since
+        // thread instance cannot provide the related data when it is not currently configured.
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:73]:
```suggestion
        // thread instance cannot provide the relevant data when it is not currently configured.
```

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 7]
```diff
@@ -0,0 +1,861 @@
...
+                {
+                    routingRole = RoutingRoleEnum::kReed;
+                }
+#endif
+            }
+            else
+            {
+                routingRole = RoutingRoleEnum::kSleepyEndDevice;
+            }
+        }
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:146]: would a switch-case be clearer?

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 8]
```diff
@@ -0,0 +1,861 @@
...
+            otNeighborInfo neighInfo;
+            otNeighborInfoIterator iterator = OT_NEIGHBOR_INFO_ITERATOR_INIT;
+
+            while (otThreadGetNextNeighborInfo(otInst, &iterator, &neighInfo) == OT_ERROR_NONE)
+            {
+                Structs::NeighborTableStruct::Type neighborTable;
+                app::DataModel::Nullable<int8_t> averageRssi;
+                app::DataModel::Nullable<int8_t> lastRssi;
+
+                if (neighInfo.mAverageRssi == OT_RADIO_RSSI_INVALID)
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:201]: Suggestion to reduce the number of lines
`(neighInfo.mAverageRssi == OT_RADIO_RSSI_INVALID) ? ... : ...;`

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp [hunk 9]
```diff
@@ -0,0 +1,861 @@
...
+                {
+                    averageRssi.SetNull();
+                }
+                else
+                {
+                    // Thread average calculation already restrict mAverageRssi to be between -128 and 0
+                    averageRssi.SetNonNull(neighInfo.mAverageRssi);
+                }
+
+                if (neighInfo.mLastRssi == OT_RADIO_RSSI_INVALID)
```
Comment (mkardous-silabs) [inline src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:211]: same here

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp (hunk 7 above)
Comment (jmartinez-silabs) [reply src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:146]: I rather not make a switch case in a switch case. This section of code was already like this in the older function

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp (hunk 5 above)
Comment (jmartinez-silabs) [reply src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:64]: We wouldn't want to compile the 2 of the 3 extra functions so splitting it in 3 would be just adding more #ifdef to the code. 

Two of those functions don't have much logic. Return error. and the other encode the correct default value. 

Not convinced splitting it is a better option

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp (hunk 8 above)
Comment (jmartinez-silabs) [reply src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:201]: cannot use .setNull in this format
compiler throws
` error: second operand to the conditional operator is of type 'void', but the third operand is neither a throw-expression nor of type 'void'`

will keep it as is

File: src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp (hunk 9 above)
Comment (jmartinez-silabs) [reply src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp:211]: same comment. can't use the `? :` conditional format

File: src/platform/BUILD.gn (hunk 4 above)
Comment (jmartinez-silabs) [reply src/platform/BUILD.gn]: done

**Review by mkardous-silabs — APPROVED** [review]

**Review by rcasallas-silabs — APPROVED** [review]

**Pull request merged**